storage/innobase/ibuf/ibuf0ibuf.cc

```diff
@@ -2389,26 +2393,25 @@ static void ibuf_read_merge_pages(const uint32_t* space_ids,
 		if (UNIV_LIKELY(page_nos[i] < size)) {
 			mtr.start();
 			dberr_t err;
-			buf_block_t *block =
+			/* Load the page and apply change buffers. */
+			std::ignore =
 			buf_page_get_gen(page_id_t(space_id, page_nos[i]),
 					 zip_size, RW_X_LATCH, nullptr,
 					 BUF_GET_POSSIBLY_FREED,
 					 __FILE__, __LINE__, &mtr, &err, true);
-			bool remove = !block
-				|| fil_page_get_type(block->frame)
-				!= FIL_PAGE_INDEX
-				|| !page_is_leaf(block->frame);
 			mtr.commit();
 			if (err == DB_TABLESPACE_DELETED) {
 				goto tablespace_deleted;
 			}
-			if (!remove) {
-				continue;
-			}
 		}
 
-		if (srv_shutdown_state == SRV_SHUTDOWN_NONE
-		    || srv_fast_shutdown) {
+		/* During slow shutdown cleanup, we apply all pending IBUF
+		changes and need to cleanup any left-over IBUF records. There
+		are a few cases when the changes are already discarded and IBUF
+		bitmap is cleaned but we miss to delete the record. Even after
+		the issues are fixed, we need to keep this safety measure for
+		upgraded DBs with such left over records. */
+		if (!slow_shutdown_cleanup) {
```

Comment:
Which cases is this referring to? Which function is cleaning the bitmap flag without deleting (or delete-marking) the change buffer record in the same mini-transaction?

Reply:
1. ibuf_delete_recs - the current bug scenario.
2. I am sure there were other cases where we had left over entries. I am not introducing this "if" condition rather replacing it. If you see a bit down you would see more comments.

Reply:
Specifically for the bug scenario it was the place below. If you want more details, please follow the steps in MDEV to repeat the issue.

ibuf_merge_or_delete_for_page()
```
                if (bitmap_bits && fseg_page_is_free(
                                space, page_id.page_no())) {
                        ibuf_mtr_start(&mtr);
                        mtr.set_named_space(space);
                        ibuf_reset_bitmap(block, page_id, zip_size, &mtr);
                        ibuf_mtr_commit(&mtr);
                        bitmap_bits = 0;
                        if (!block
                            || btr_page_get_index_id(block->frame)
                            != DICT_IBUF_ID_MIN + IBUF_SPACE_ID) {
                                ibuf_delete_recs(page_id);
                        }
                }

```